### PR TITLE
Feat(medusa): Convert middleware service to typescript

### PR DESCRIPTION
**Pull request by pKorsholm**
Fixes CORE-351

### CODE REVIEW

File: packages/medusa/src/services/middleware.ts [hunk 1]
```diff
@@ -1,31 +1,51 @@
...
+  middleware: (options: Record<string, unknown>) => RequestHandler
+  options: Record<string, unknown>
+}
 
 /**
  * Orchestrates dynamic middleware registered through the Medusa Middleware API
  */
-class MiddlewareService {
-  constructor(container) {
+class MiddlewareService extends TransactionBaseService<MiddlewareService> {
```
Comment (srindom) [inline packages/medusa/src/services/middleware.ts:14]: todo: shouldn't be necessary to convert this to a transactions service as it doesn't interact with the db layer

**Review by srindom — CHANGES_REQUESTED** [review]
Small fix here

File: packages/medusa/src/services/middleware.ts (hunk 1 above)
Comment (adrien2p) [reply packages/medusa/src/services/middleware.ts:14]: indeed, this service does not need that extension.

File: packages/medusa/src/services/middleware.ts [hunk 2]
```diff
@@ -1,31 +1,51 @@
...
+  protected transactionManager_: EntityManager | undefined
+
+  protected readonly postAuthentication_: middlewareType[]
+  protected readonly preAuthentication_: middlewareType[]
+  protected readonly preCartCreation_: RequestHandler[]
+  protected readonly routers: Record<string, Router[]>
+
+  constructor() {
+    // eslint-disable-next-line prefer-rest-params
+    super(arguments[0])
```
Comment (adrien2p) [inline packages/medusa/src/services/middleware.ts:25]: todo: nothing to pass here

File: packages/medusa/src/services/middleware.ts [hunk 3]
```diff
@@ -1,31 +1,51 @@
...
-  addRouter(path, router) {
+  addRouter(path: string, router: Router): void {
     const existing = this.routers[path] || []
     this.routers[path] = [...existing, router]
   }
 
-  getRouters(path) {
+  getRouters(path: string): Router[] {
     const routers = this.routers[path] || []
     return routers
```
Comment (adrien2p) [inline packages/medusa/src/services/middleware.ts:40]: suggestion: we can directly return `this.routers[path] || []`

File: packages/medusa/src/services/middleware.ts [hunk 4]
```diff
@@ -1,31 +1,51 @@
...
   }
 
   /**
    * Validates a middleware function, throws if fn is not of type function.
    * @param {function} fn - the middleware function to validate.
+   * @returns nothing if the middleware is a function
    */
-  validateMiddleware_(fn) {
+  validateMiddleware_(fn): void {
```
Comment (adrien2p) [inline packages/medusa/src/services/middleware.ts:48]: todo: Missing typings on `fn`

File: packages/medusa/src/services/middleware.ts [hunk 5]
```diff
@@ -40,9 +60,12 @@ class MiddlewareService {
    *   middleware function.
    * @param {object} options - the arguments that will be passed to the
    *   middleware
-   * @return {void}
+   * @return void
    */
-  addPostAuthentication(middleware, options) {
+  addPostAuthentication(
+    middleware: (options: Record<string, unknown>) => RequestHandler,
```
Comment (adrien2p) [inline packages/medusa/src/services/middleware.ts:66]: suggestion: In this service at the top, you could create a local type such as 
```ts
type MiddlewareHandler = (options: Record<string, unknown>) => RequestHandler
```

File: packages/medusa/src/services/middleware.ts (hunk 1 above)
Comment (pKorsholm) [reply packages/medusa/src/services/middleware.ts:14]: I was on the fence but included it for consistency, but I will remove it 😄

**Review by srindom — APPROVED** [review]
LGTM!

**Pull request merged**